crates/proof-of-sql/src/sql/proof_exprs/range_check.rs

```diff
@@ -0,0 +1,137 @@
+        let scalar_bytes = bytemuck::cast_slice::<u64, u8>(&scalar_array); // Safer casting using bytemuck
+
+        // Populate columns and update byte counts
+        for (col, &byte) in columns.iter_mut().zip(scalar_bytes.iter()) {
+            col[i] = byte;
+
+            // Update the byte count in the corresponding position
+            byte_counts[byte as usize] += S::one(); // Increment the count of the byte value
+        }
+    }
+
+    // Allocate and initialize byte_values to represent each possible byte as a scalar directly
+    let byte_values: &mut [S] =
+        alloc.alloc_slice_fill_with(256, |i| S::try_from(i.into()).unwrap());
+
+    // 1. Produce an MLE over each column of words
+    for column in columns {
+        builder.produce_intermediate_mle(column as &[u8]);
+    }
+
+    // 2. Produce the anchored MLE that the verifier has access to, consisting
+    // of all possible word values. These serve as values to lookup
+    // in the lookup table
+    builder.produce_anchored_mle(byte_values as &[S]);
+
+    // 3. Next produce an MLE over the counts of each word value
+    builder.produce_intermediate_mle(byte_counts as &[S]);
+
+    // Invert the scalars, and get the inverted words.
+    // This modifies the column in place.
+    slice_ops::batch_inversion(&mut scalars[..]);
+    let mut inverted_word_columns: Vec<&mut [S]> = (0..31)
+        .map(|_| alloc.alloc_slice_fill_with(scalars.len(), |_| S::ZERO))
+        .collect();
+
+    // Get the alpha challenge from the verifier
+    let alpha = builder.consume_post_result_challenge();
+
+    // Iterate through the inverted scalars and fill columns
+    for (i, inverted_scalar) in scalars.iter().enumerate() {
+        let inverted_scalar_array: [u64; 4] = (*inverted_scalar).into();
+        let inverted_scalar_words = bytemuck::cast_slice::<u64, u8>(&inverted_scalar_array);
+
+        // Allocate and initialize row for each inverted scalar processing
+        let row: &mut [S] = alloc.alloc_slice_fill_with(inverted_scalar_words.len(), |_| S::zero());
+
+        for ((col, &inverted_word), row_entry) in inverted_word_columns
+            .iter_mut()
+            .zip(inverted_scalar_words.iter())
+            .zip(row.iter_mut())
+        {
+            // Convert a word into a scalar so that we can perform arithmetic on it
+            let value =
+                S::try_from(inverted_word.into()).expect("u8 will always fit in scalar") + alpha;
+            col[i] = value;
+            *row_entry = value;
+        }
+        builder.produce_intermediate_mle(&*row);
+    }
+
+    // Now produce an intermediate MLE over the inverted word values + verifier challenge alpha
+    let inverted_word_values: &mut [S] =
+        alloc.alloc_slice_fill_with(256, |i| S::try_from(i.into()).unwrap() + alpha);
+    slice_ops::batch_inversion(&mut inverted_word_values[..]);
+    builder.produce_anchored_mle(inverted_word_values as &[S]);
```

Comment:
This is not anchored, since the verifier doesn't have the commitment to it (because it doesn't know alpha ahead of time.)